sycl/source/device_selector.cpp

```diff
@@ -51,15 +41,22 @@ static bool isForcedDevice(const device &Dev, int Index = -1) {
   else
     Backend = detail::getSyclObjImpl(Dev)->getPlugin().getBackend();
 
-  for (const detail::device_filter &Filter : FilterList->get()) {
-    if ((Filter.Backend == Backend || Filter.Backend == backend::all) &&
-        (Filter.DeviceType == Type ||
-         Filter.DeviceType == info::device_type::all)) {
-      if (Index < 0 || (Filter.HasDeviceNum && Filter.DeviceNum == Index))
-        return true;
-    }
-  }
-  return false;
+  std::vector<cl::sycl::device> AllDevices =
+      cl::sycl::detail::getSyclObjImpl(Dev.get_platform())
+          ->get_devices(info::device_type::all, true);
+  int Index = std::distance(
+      AllDevices.begin(), std::find(AllDevices.begin(), AllDevices.end(), Dev));
```

Comment:
It seems we did not make it clear the definition of device number.
Device number is a global number in front of the device name at each line printed by sycl-ls without --verbose.
The device number is not a local number within each platform.
Here is an example output from sycl-ls tool.
> 1. ACC : Intel(R) FPGA Emulation Platform for OpenCL(TM) 1.2 [2020.11.11.0.04_160000]
> 2. GPU : Intel(R) OpenCL HD Graphics 3.0 [20.44.18297]
> 3. CPU : Intel(R) OpenCL 2.1 [2020.11.11.0.04_160000]
> 4. GPU : Intel(R) Level-Zero 1.0 [1.0.18297]
> 5. HOST: SYCL host platform 1.2 [1.2]
`
Please note that SYCL_DEVICE_NUMBER=1 would be vague if you treat '1' as the local index within each platform.
@romanovvlad @smaslov-intel , what do you think?

Reply:
This is misunderstanding form my perspective. SYCL_DEVICE_FILTER always operated with device IDs within platform.
That was before my change and it stays the same after. if this behavior need to be changed it should be done as separate PR.

Reply:
No, that's not ture. My original implementation calculated the device index correctly across platforms.
```
device device_selector::select_device() const {
  vector_class<device> devices = device::get_devices();
  ...
  for (const auto &dev : devices) {
   ...
    // If SYCL_DEVICE_FILTER is set, give a bonus point for the device
    // whose index matches with desired device number.
    int index = &dev - &devices[0];
    if (isForcedDevice(dev, index)) {
      dev_score += 1000;
    }
```

Reply:
That conflicts with documentation: "Device_num is an integer that indexes the enumeration of devices from the *sycl::platform::get_device()* call, where the first device in that enumeration has index zero.  " (see https://github.com/intel/llvm/blob/sycl/sycl/doc/EnvironmentVariables.md).
And finally there bugs in that implementation because:
 - device_num field only affects device_selector() other selectors ignore it.
 - custom_selectors/user_selectros will ignore SYCL_DEVICE_FILTER.

Reply:
Hmm..
I wrote that document based on my original implementation.
The original implementation calls get_devices() and find the index correctly.
I don't understand what conflicts/bug you are referring to.
All device selectors are affected by SYCL_DEVICE_FILTER.
Let me repeat, SYCL_DEVICE_FILTER is dual-purposed, and it already filter out irrelevant plugins and not even load it, which means no related platforms and devices are included in the enumeration returned from get_devices().
For example, SYCL_DEVICE_FILTER=level_zero, no opencl platforms or devices are available because opencl plugin is not even loaded.
custom selectors cannot ignore SYCL_DEVICE_FILTER for this reason, too.